etcd3locker/locker.go

```diff
@@ -0,0 +1,120 @@
+// Tested on etcd 3.1/3.2/3.3
+package etcd3locker
+
+import (
+	"context"
+	"errors"
+	"log"
+	"strings"
+	"sync"
+
+	etcd3 "github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/tus/tusd"
+)
+
+var (
+	ErrLockNotHeld = errors.New("Lock not held")
+	DefaultTtl     = int64(60)
+)
+
+type Etcd3Locker struct {
+	// etcd3 client session
+	Client *etcd3.Client
+
+	// locks is used for storing Etcd3Locks before they are
+	// unlocked. If you want to release a lock, you need the same locker
+	// instance and therefore we need to save them temporarily.
+	locks  map[string]*Etcd3Lock
+	mutex  sync.Mutex
+	prefix string
+}
+
+// New constructs a new locker using the provided client.
+func New(client *etcd3.Client) (*Etcd3Locker, error) {
```

Comment:
Would it be possible to require `client` to be an interface here instead of `etcd3.Client`. The problem is that if we use the type `etcd3.Client` we cannot add etcd3's packages to the vendor directory. We already have to deal with this in the s3store and consullocker which is the reason why their dependencies are not vendored but I would like to avoid this issue in the future.

Reply:
Extracting the `client` to be an interface doesn't seem like it would get us away from having to vendor etcd3's packages (as I understand it). I don't have a concrete idea of how this idea would look like - but it seems like it might make things a bit more confusing to understand.  Some clarification would be helpful here so I have a better idea of what the end goal looks like. 

The batteries-included tusd can be a dependency nightmare (e.g. having to manage gcs/s3/consul/etcd3 deps) such that I wonder if it might make sense to have these 'external service plugins' be their own repos similar to the way how Terraform extracted their supported providers into separate bundles that people pull in (as they require it). 

e.g. If I need s3store and etcd3locker - I would pull in the following packages and glue together the proper components as such: 

```
import (
   s3store  "github.com/tus/tusd-s3-store"
   etcd3Locker "github.com/tus/tusd-etcd3-locker"
)

...

storage := s3store.New(..)
locker := etcd3Locker.New(..)
composer := tusd.NewStoreComposer()
storage.UseIn(composer)
locker.UseIn(composer)
```

This would keep `tusd` easy to build and free from having to do weird Travis CI things as more 'external service plugins` are added. It's possible that people who are pulling in these plugins (and intend to use them) have more of a sense of how to manage the dependencies they intend to use rather than make everyone pay that cost.

Reply:
My comment was a bit confusion, apologies for that. The problem is that if we expose a function whose signature contains a type from a different package we must not vendor this different package (see https://github.com/etcd-io/etcd/issues/4913#issuecomment-204298576). We had problems with this in the past with the AWS Go SDK (https://github.com/tus/tusd/issues/59).

However, since you didn't vendor `github.com/coreos/etcd` , this is not a problem anymore. You did everything correct, please ignore my original comment.

In the future I may try to even get rid of vendoring in the tusd library at all, since libraries should not vendor their dependencies (https://peter.bourgon.org/go-best-practices-2016/#dependency-management). A bit of topic, but this is not as easy as it seems since tusd is both a library and binary. etcd had a similar problem and they did have to work around it to some extend (https://github.com/etcd-io/etcd/issues/4913).

> The batteries-included tusd can be a dependency nightmare (e.g. having to manage gcs/s3/consul/etcd3 deps) such that I wonder if it might make sense to have these 'external service plugins' be their own repos similar to the way how Terraform extracted their supported providers into separate bundles that people pull in (as they require it).

What exactly would be the benefit of having them in separate repositories than in separate packages as it is right now? I am not sure how exactly `go get` behaves but does it require the AWS SDK through `github.com/tus/tusd/s3store` if I am only using `github.com/tus/tusd/gcsstore`?

Reply:
> What exactly would be the benefit of having them in separate repositories than in separate packages as it is right now? I am not sure how exactly go get behaves but does it require the AWS SDK through github.com/tus/tusd/s3store if I am only using github.com/tus/tusd/gcsstore?

I'll need to double-check as I might've had the misconception that doing `go get github.com/tus/tusd` would require all subsequent deps (within the package) to be properly satisfied. I believe you're correct in that referencing the specific packages such as `github.com/tus/tusd/s3store` should only require the AWS SDK and not the `github.com/tus/tusd/gcsstore`. 

That being said, building the batteries-included `tusd` through `cmd/tusd/main.go` if we want to make `gcsstore`/`s3store`/`consullocker`/`etcd3locker`/`more` seems like it may become a pain with getting the right dependencies and their versions.

We're currently using `dep` in our current environment (which makes things pretty easy for us), but I'm not sure what the direction of the pre-built `tusd` binary in this repository is going to go if it seems painful to build the binary from source unless the prebuilt `tusd` binary is not going to include either the etcd3locker/consullocker to keep things simple.

Reply:
> I'll need to double-check as I might've had the misconception that doing go get github.com/tus/tusd would require all subsequent deps (within the package) to be properly satisfied.

As I said, I am also not sure if my statement was correct but we will see.

> That being said, building the batteries-included tusd through cmd/tusd/main.go if we want to make gcsstore/s3store/consullocker/etcd3locker/more seems like it may become a pain with getting the right dependencies and their versions.

I am not sure about that. In the past we didn't have much problems with getting the correct versions for our dependencies. Since our dependencies on the AWS SDK, Google Cloud SDK etc are rather distinct, version conflicts should not happen.
However, I agree that the feature of the tusd binary will increase as we add more packages but for me this is the entire idea of the pre-built binary. It makes it easy for users to get started with tusd and takes the burden of compiling it on your own (for example, at Transloadit we use the pre-built binary in production). So our plan is definitely to support this pre-built binary in the future although it might not support every since package.

Reply:
It's entirely possible that our troubles are stemming from the fact that `etcd3` master branch can be in a broken state 😢 (clearly from the amount of trouble I've had here). Glad to know that the other SDKs aren't being as troublesome if that is the case.

Reply:
I'm sorry to hear that! Hopefully, we will be able to pin the version using gopkg.in to avoid unstable dependencies.